producer/buffer/buffer.go

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package buffer
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/producer/data"
+
+	"github.com/uber-go/atomic"
+	"github.com/uber-go/tally"
+)
+
+var (
+	errBufferFull                         = errors.New("buffer full")
+	errBufferClosed                       = errors.New("buffer closed")
+	errInvalidDataLargerThanMaxBufferSize = errors.New("invalid data, larger than max buffer size")
+)
+
+type bufferMetrics struct {
+	messageDropped  tally.Counter
+	bytesDropped    tally.Counter
+	messageBuffered tally.Gauge
+	bytesBuffered   tally.Gauge
+}
+
+func newBufferMetrics(scope tally.Scope) *bufferMetrics {
+	return &bufferMetrics{
+		messageDropped:  scope.Counter("message-dropped"),
+		bytesDropped:    scope.Counter("bytes-dropped"),
+		messageBuffered: scope.Gauge("message-buffered"),
+		bytesBuffered:   scope.Gauge("bytes-buffered"),
+	}
+}
+
+type buffer struct {
+	sync.Mutex
+
+	strategy        OnFullStrategy
+	maxBufferSize   uint64
```

Comment:
Any particular reason to use a uint64 here? It's more common to use an int for stuff like len or size in Go.

Reply:
it makes the calculation here easier since the buffer size is an atomic uint64, could change the type exposed in options to int.